WireDomain/Sources/WireDomain/NotificationService/NotificationService.swift

```diff
@@ -0,0 +1,177 @@
+    // MARK: - Object lifecycle
+
+    override init() {
+        logger.info("initializing new legacy notification service")
+        super.init()
+    }
+
+    // MARK: - Notifications
+
+    override func didReceive(
+        _ request: UNNotificationRequest,
+        withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
+    ) {
+        ongoingTask?.cancel()
+        let cookieStorage: ZMPersistentCookieStorage = Injector.resolve()
+        let isAuthenticated = cookieStorage.isAuthenticated
+
+        guard isAuthenticated else {
+            logger.error(
+                "Not displaying notification because app is not authenticated"
+            )
+
+            return finishWithEmptyNotification()
+        }
+
+        self.contentHandler = contentHandler
+
+        ongoingTask = Task {
+            do {
+                let notificationUserInfo = request.content.userInfo
+
+                let notification = try NotificationPayload(
+                    userInfo: notificationUserInfo
+                )
+
+                notificationSession = try await createNotificationSession(
+                    userID: notification.userID
+                )
+
+                try await notificationSession?.processPushNotification(
+                    eventID: notification.eventID
+                )
+
+            } catch {
+                logError(error)
+                finishWithEmptyNotification()
+            }
+        }
+    }
+
+    override func serviceExtensionTimeWillExpire() {
+        logger.warn("legacy service extension will expire")
+        finishWithEmptyNotification()
+    }
+
+    private func createNotificationSession(
+        userID: UUID
+    ) async throws -> NotificationSession {
+        let userLocalStore: UserLocalStoreProtocol = Injector.resolve()
+        let selfUserInfo = await userLocalStore.selfUserInfo()
+
+        guard let selfClientID = selfUserInfo.clientId else {
+            throw Failure.missingSelfClientID
+        }
+
+        let updateEventsRepository = UpdateEventsRepository(
+            userID: userID,
+            selfClientID: selfClientID,
+            // these were already initialized, resolving them
+            updateEventsAPI: Injector.resolve(),
+            pushChannel: Injector.resolve(),
+            updateEventDecryptor: Injector.resolve(),
+            updateEventsLocalStore: Injector.resolve()
+        )
+
+        return NotificationSession(
+            updateEventsRepository: updateEventsRepository
+        ) { [weak self] notificationContent in
+            self?.finishWithNotification(content: notificationContent)
+        }
+    }
+
+    private func finishWithNotification(content: UNNotificationContent) {
+        // TODO: [WPB-11175]
+    }
+
+    private func finishWithEmptyNotification() {
+        logger.info("finishing without showing notification")
+        let emptyNotification = UNNotificationContent()
+        contentHandler?(emptyNotification)
```

Comment:
```suggestion
    // With the "filtering" entitlement, we can tell iOS to not display a user notification by
    // passing empty content to the content handler.
    // See https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering
        contentHandler?(emptyNotification)
```

Adding some comments here